main.cpp

```diff
@@ -318,460 +267,310 @@ struct HeapA
  Wait for my code review.
  */
 
-
-struct FloatType
+template<typename T>
+struct Numeric
 {
-    explicit FloatType(float f) : value(new float(f)) {}
-    ~FloatType()
-    {
-        delete value;
-    } 
-
-    FloatType& operator+=(float lhs);
-    FloatType& operator-=(float lhs);
-    FloatType& operator*=(float lhs);
-    FloatType& operator/=(float lhs);
+    using Type = T;
 
-    FloatType& pow(FloatType&);
-    FloatType& pow(DoubleType&);
-    FloatType& pow(IntType&);
-    FloatType& pow(float);
+    Numeric(Type num) : value(std::make_unique<Type>(num)) {}
+    ~Numeric() {}
 
-    FloatType& apply(std::function<FloatType&(float&)> func);
-    FloatType& apply(void(*func)(float&));
-
-    operator float() const
+    Numeric& operator+=(Type rhs) 
     {
-        return *value;
+        *value += rhs;
+        return *this;
     }
     
-
-private:
-    float* value;
-    FloatType& powInternal(const float);
-};
-
-struct DoubleType
-{
-    explicit DoubleType(double d) : value(new double(d)) {}
-    ~DoubleType()
+    
+    Numeric& operator-=(Type rhs)
+    {       
+        *value -= rhs;
+        return *this;  
+    }
+    
+    
+    Numeric& operator*=(Type rhs)
     {
-        delete value;
-    } 
-
-    DoubleType& operator+=(double lhs);
-    DoubleType& operator-=(double lhs);
-    DoubleType& operator*=(double lhs);
-    DoubleType& operator/=(double lhs);
-
+        *value *= rhs;
+        return *this;
+    }
+    
+    
+    template<typename P>
+    Numeric& operator/=(P rhs)
```

Comment:
Help yourself and use names that tell you what something is.
`P` doesn't reveal any information to you, the programmer, when you try to use this `/=` function.
Consider how much information you receive as the developer with the following template declaration:
```
template<typename OtherType>
Numeric& operator /= (const OtherType& rhs)
```
The amount of information you receive just from using a better parameter type name is why we have the rule:
`Name things what they are`
- [ ] revise